lib/streamlit/components/v1/components.py

```diff
@@ -109,6 +108,33 @@ def create_instance(
         if len(args) > 0:
             raise MarshallComponentException(f"Argument '{args[0]}' needs a label")
 
+        try:
+            import pyarrow
+            from streamlit.elements import arrow_table
+        except ImportError:
+            import sys
+
+            if sys.version_info >= (3, 9):
+                raise StreamlitAPIException(
+                    """To use Custom Components in Streamlit, you need to install
+PyArrow. Unfortunately, PyArrow does not yet support Python 3.9.
+
+You can either switch to Python 3.8 with an environment manager like PyEnv, or
+[install Streamlit with conda](https://discuss.streamlit.io/t/note-installation-issues-with-python-3-9-and-streamlit/6946/7):
+
+`conda install -c conda-forge streamlit`
+"""
+                )
+            else:
+                raise StreamlitAPIException(
+                    """To use Custom Components in Streamlit, you need to install
+PyArrow. To do so locally:
+
+`pip install pyarrow`
+
+And if you're using Streamlit Sharing, add "pyarrow" to your requirements.txt."""
+                )
+
```

Comment:
probably not worth it but just noting this down, it looks all custom components will not work regardless or not if the component uses tables.

Reply:
Good point. The alternate approach would be to try-except specifically the areas that call on arrow_table, or to move this try-catch into `arrow_table.py` altogether.

In my head, cutting the boundary at "Custom Components" rather than "Custom Components using tables" was easier for users to reason about, and I imagine those who run into this issue (those inclined to install components) are technical enough not to have a huge problem with this, plus AFAICT components are not that widely, so I didn't feel like optimizing this further.

Incidentally, would be nice to have easy metrics on "What % of Streamlit Apps use Components" and "What % of Streamlit Apps are on Python 3.9".